drivers/dp/swdp_ll_pin.h

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/gpio.h>
+
+#if defined(CONFIG_SOC_SERIES_NRF52X)
+#define CPU_CLOCK		64000000U
+#else
+#define CPU_CLOCK		CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC
+#endif
+
+#if defined(CONFIG_SOC_SERIES_NRF52X) || defined(CONFIG_SOC_SERIES_NRF53X)
+#define FAST_BITBANG_HW_SUPPORT 1
+#else
+#define FAST_BITBANG_HW_SUPPORT 0
+#endif
+
+static ALWAYS_INLINE void pin_delay_asm(uint32_t delay)
+{
+#if defined(CONFIG_CPU_CORTEX_M)
+	__asm volatile ("movs r3, %[p]\n"
+			".start_%=:\n"
+			"subs r3, #1\n"
+			"bne .start_%=\n"
+			:
+			: [p] "r" (delay)
+			: "r3", "cc"
+			);
+#else
+#warning "Pin delay is not defined"
+#endif
+}
+
+static ALWAYS_INLINE void swdp_ll_pin_input(void *const base, uint8_t pin)
+{
+#if defined(CONFIG_SOC_SERIES_NRF52X) || defined(CONFIG_SOC_SERIES_NRF53X)
+	NRF_GPIO_Type * reg = base;
+
+	reg->PIN_CNF[pin] = 0b0000;
```

Comment:
HAL should be used here, and in code below.

Reply:
this is from upstream zephyr - it's not the right place to change it here.

Reply:
It should be fixed in a follow-up PR in upstream, that would be cherry-picked as fromlist to this PR.

Reply:
@masz-nordic Could you give me some pointers on how to do that? `nrf_gpio_cfg` uses quite a few more cycles then just setting the PIN_CNF register directly. Speed is of utmost importance for this part.

Reply:
Could you post disassembly here? Is `nrf_gpio_pin_port_decode` not getting optimized enough?
We can add new functions if needed.